### DISCUSSION TITLE
Use postgresql@15 instead of deprecated postgresql

### DISCUSSION THREAD

**Discussion by theory** [Homebrew/discussions#4391] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: 931327df1f30a3a5b1aa07657c0d1070e61988f9
Last commit: 8 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: cf43a01aa352a63fd3a326a020143fd0650e96f4
Core tap last commit: 7 weeks ago
Core tap branch: master
Core tap JSON: 08 Apr 16:27 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: emacs
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  sqitch
```

### Description of issue

[This Sqitch Tap issue](/sqitchers/homebrew-sqitch/issues/56), opened this week, reports that plain old `postgresql` is deprecated as a dependency. Can confirm!

```
$ brew services stop postgresql   
Warning: Use postgresql@14 instead of deprecated postgresql
Warning: Service `postgresql@14` is not started.
```

The dependency is set via `--with-postgres-support` and configured [in the Sqitch tap formula](/sqitchers/homebrew-sqitch/blob/04621f0/Formula/sqitch.rb#L36-L39):

``` ruby
  if build.with? "postgres-support"
    depends_on 'postgresql' => :recommended
    depends_on 'libpq' => :build
  end
```

Wondering what I should use as a dependency instead. Should it specify `@15` and tell people to manually install it if they want another version? Is there no `@latest`? Does this apply to other dependencies (Sqitch supports MySQL, SQLite, and a bunch of other [options](https://github.com/sqitchers/homebrew-sqitch/blob/04621f0/Formula/sqitch.rb#L17-L25). What's the recommended pattern here?

Thanks,

David

**Comment by ZhongRuoyu** [comment, marked as answer]
> Is there no `@latest`?

No, there is no "latest" alias for `postgresql`. It was Homebrew's decision to make `postgresql` a version-only formula, because "it breaks on every major version upgrade until people run a tool we provide for them which is not 100% bug-free" (https://github.com/Homebrew/homebrew-core/pull/107726). Here "version-only" means that the `postgresql` alias has been deprecated, and users should specify the version of `postgresql` to install, like `postgresql@14`, `postgresql@15`, etc. See also: https://github.com/orgs/Homebrew/discussions/4318#discussioncomment-5275476.

> Does this apply to other dependencies

*Usually*, for other formulae like `llvm` and `openssl`, the one without the `@version` suffix is either the latest (`llvm`) or an alias for the latest (`openssl`; at the time of writing, it is an alias for `openssl@3`).

Sometimes, a non-versioned alias for a versioned formula may only point to the latest version after some delay. This has been the case for `python` -- there is a delay between when a new `python` version is made available and when the `python` alias is switched to that latest version, because we'd like most of the dependants to switch to that new release first.

Also, do note that there is only one version of [`sqlite`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/sqlite.rb) in `Homebrew/core`. (Since you mentioned `sqlite`.)

> Wondering what I should use as a dependency instead.

Tap maintainers are advised to use the canonical names for dependencies, and only migrate to a newer version after it has been tested. (Using canonical names is a must in `Homebrew/core`.) Often, formula upgrades are not completely backwards compatible.

The canonical name of a formula is its filename without the `.rb` extension. For instance, [`llvm`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/llvm.rb) and [`openssl@3`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/openssl@3.rb) are canonical names.

> Should it specify `@15` and tell people to manually install it if they want another version?

I'm not able to give specific advice on which `postgresql` version to use. But since the [`sqitchers/sqitch/sqitch`](https://github.com/sqitchers/homebrew-sqitch/blob/main/Formula/sqitch.rb) formula is already using `option`s, one possible approach is to consider supporting multiple versions as different options, like `option "with-postgres14-support"`, etc.

**Reply by theory** [reply]
Thanks for the reply, @ZhongRuoyu.

>  It was Homebrew's decision to make postgresql a version-only formula, because "it breaks on every major version upgrade until people run a tool we provide for them which is not 100% bug-free"

That makes sense. Makes me think as an app developer, though, that it'd be handy to have `psql` and `libpq-devel` type packages that aren't as subject to such issues. Sqitch for example, depends on and uses `psql` and the libpq library to compile drivers.

> *Usually,* for other formulae like `llvm` and `openssl`, the one without the `@version` suffix is either the latest (`llvm`) or an alias for the latest.

Great, that makes sense. Looks like [MySQL](https://github.com/Homebrew/homebrew-core/blob/master/Formula/mysql.rb) doesn't require the `@` version.

> Tap maintainers are advised to use the canonical names for dependencies. 

Looks like I should update the OpenSSL dependency too, then.

>  one possible approach is to consider supporting multiple versions as different options, like `option "with-postgres14-support"`, etc.

Yeah, then would just have to stay on top of the new releases every year. Not a big deal, though the ability to just require `psql` and libpq would be handy. Though I guess some would still want to install Postgres to test locally.

**Reply by ZhongRuoyu** [reply]
> it'd be handy to have `psql` and `libpq-devel` type packages that aren't as subject to such issues

Yea, probably. Though one problem is: I can't think of a way to create a `psql` formula that doesn't conflict with `postgresql@14` (the "primary" `postgresql` that's not keg-only). If we split out `psql`, then that would probably hurt the experience of existing users; if we keep `postgresql` formulae as they are and simply create a `psql` for the CLI client, then it'll need to be keg-only (because it conflicts with `postgresql@14`), and thus not quite useful.

On a side note, there is a [`libpq`](https://formulae.brew.sh/formula/libpq) formula available.

**Reply by theory** [reply]
Yeah, it could be fussy. Does the libpq formula include header and library files?

**Reply by theory** [reply]
Oh, [looks like it](https://github.com/Homebrew/homebrew-core/blob/f0a4d69/Formula/libpq.rb#L44-L50), nice!

**Reply by ZhongRuoyu** [reply]
Yes it does. Actually, let me correct myself: `libpq` *does* provide `psql` among other binaries. But it's keg-only (because, as I mentioned earlier, it conflicts with `postgresql` formulae), so `$(brew --prefix libpq)/bin` needs to be in `PATH`.

**Reply by theory** [reply]
I see. I wonder if it would make sense to add some sort of option for that instead of the full Postgres server (or in addition to it) and add it to the path when Sqitch runs or something. Do the libraries and includes need to be added to environment variables to build against, too?

**Reply by ZhongRuoyu** [reply]
> Do the libraries and includes need to be added to environment variables to build against, too?

In fact, environment variables that affect libraries and includes, namely `LDFLAGS` and `CPPFLAGS`, should already be set for the dependencies. Even for keg-only ones.

> add it to the path when Sqitch runs or something

If you want to make `psql` available to Sqitch, you may want to use `write_env_script`. [Here's an example](https://github.com/Homebrew/homebrew-core/blob/master/Formula/check_postgres.rb#L35) from the `check_postgres` formula.

Also, there's an `env_script_all_files` method. Example [here](https://github.com/Homebrew/homebrew-core/blob/ca990de54218519966bf67071d866dd1025a0cd3/Formula/solr.rb#L34).

**Reply by theory** [reply]
So if I included a path to the libpq stuff in `LDFLAGS` and `CPPFLAGS` it should work? The formula already [appends some paths for OpenSSL support](/sqitchers/homebrew-sqitch/blob/04621f0/Formula/sqitch.rb#L87-L89).  Sorry for the dumb questions, I tweak this formula like once every five years and have to learn the Homebrew patterns all over again.

**Reply by ZhongRuoyu** [reply]
No worries. You shouldn't even need to explicitly set `LDFLAGS` and `CPPFLAGS`, actually (provided that the includes and libraries are in their respective standard locations, namely `include` and `lib`). The required environment variables are already set for you, even for keg-only formulae.

**Reply by theory** [reply]
Well then I'm confused. I have no idea what the right thing to do here is, though just chaining the requirement to `postgres@15` would be simplest.

**Reply by ZhongRuoyu** [reply]
Unless `libpq` (including libraries and executables like `psql`) is not sufficient, I think `postgres` may not actually be needed. If it needs `psql` during runtime, then `write_env_script` will help the executables find `psql`.

**Reply by theory** [reply]
Well then I wonder if it should prefer `libpq` and `mysql-client`, then:

``` diff
--- a/Formula/sqitch.rb
+++ b/Formula/sqitch.rb
@@ -34,7 +34,7 @@ class Sqitch < Formula
   end
 
   if build.with? "postgres-support"
-    depends_on 'postgresql' => :recommended
+    depends_on 'libpq' => :recommended
     depends_on 'libpq' => :build
   end
 
@@ -43,7 +43,7 @@ class Sqitch < Formula
   end
 
   if build.with? "mysql-support"
-    depends_on 'mysql' => :recommended
+    depends_on 'mysql-client' => :recommended
   end
 
   if build.with? "firebird-support"
```

This would get it out of the business of installing a full service.

**Reply by theory** [reply]
Will need to futz with `write_env_script` for both I would guess. Perhaps an array to which it appends `Formula["libpq"].opt_bin` when building for Postgres support and `Formula["mysql-client"].opt_bin` when building for MySQL support. Then after building everything, move the `sqitch` command and wrap it with `write_env_script` if anything is in the array.

**Reply by ZhongRuoyu** [reply]
> Well then I wonder if it should prefer `libpq` and `mysql-client`, then:

Yes, I think this is the direction, if the services are not required.

**Reply by theory** [reply]
@ZhongRuoyu would you mind skimming sqitchers/homebrew-sqitch#57 to makes sure I haven't done anything too dumb? I tested it on my Mac and properly finds the clients.
